### ISSUE TITLE
Fake Sig Protection

### ISSUE THREAD

**Issue by robsee** [microg/android_packages_apps_GmsCore#243]
Am I correct that any app can take advantage of the fake sig support if it is compiled into a rom ? How hard would it be for the patches to be modified to require microg to be signed with a key held by the rom builder ? Would this fully protect from a malicious app doing bad stuff using fake sig support ?

**Comment by Lanchon** [comment]
it wouldn't be hard at all. what bad stuff would a malicious app do? who would want an OS that doesn't let you install the microG build that you choose?

**Comment by n76** [comment]
I guess I don't see the issue with a fake signature permission: It is just another permission that the app would need to ask for. And with Marshmallow or Nougat, the user/owner of the phone would have to explicitly grant that permission. For myself, the only apps that I'd grant that permission to would be GmsCore and possibly BlankStore. Any other app that asked for that permission would be denied by me.

If the app is compiled into the ROM, then it could just as well be compiled into the privileged apps area and need not ask for permission regardless of the status of a permission patch.

**Comment by ArchangeGabriel** [comment]
AFAIK, OmniROM implemented signature faking with a permission system. The same thing that was rejected by CM devs.

**Comment by mar-v-in** [comment]
@robsee to restrict it to apps that are signed with a key only the rom builder holds, you can easily change the permission protection level to `signature`, if you want to restrict it to apps residing in the `/system/priv-app` you can set protection level to `system` (or `privileged` in newer android versions). The relevant line in the patch is [this one](https://github.com/microg/android_packages_apps_GmsCore/blob/master/patches/android_frameworks_base-M.patch#L32). It's really that simple, you don't have to consider something else (the patch is a small one anyway)

**Comment by robsee** [comment]
@mar-v-in Thanks. It would purely be for my personal build of copperheadOS. Maybe it really isn't an issue, but couldn't a bad app provide the fake sig of a good app to access its data ?

**Comment by n76** [comment]
A bad app could try to fake the sig of a good app to access its data, but you would have to give it permission to do so. So if you don't give it permission to fake a signature you are safe from that particular problem.  At least that is how it works on Marshmallow (I haven't migrated to Nougat yet).

**Comment by mar-v-in** [comment]
If an app can access data by using signature spoofing, this is only because another third-party app uses an unusual way to protect data. Using signatures directly is not intended by AOSP, as mentioned in the developer reference and also a warning is shown while compiling the app if it does. A bad app can not use the private data storage of the genuine app itself, only data stored by a third party for the genuine app might become accessible.

You might want to read [this blog post](http://blogs.fsfe.org/larma/2016/microg-signature-spoofing-security/) for more details.

**Comment by Lanchon** [comment]
no, it cant. android itself doesnt believe the fake signature and the standard security model of the OS is enforced against the real app signature, not the fake one. for the OS, it is just a regular app. the only thing changed is what is reported by the OS to be the signature of the app, if the app's signature is queried by another app, and that's it.

**Comment by BryanByteZero** [comment]
A bit offtopic but which app I can use to control the fake sig permission? I use the xposed module FakeGapps to enable it. BTW running Mokee ROM.

**Comment by mar-v-in** [comment]
@BryanByteZero 
FakeGapps is not a full implementation of signature spoofing. It will just spoof all installed play store and play services like apps to be the real ones. Thus there is nothing to control.
